hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java

```diff
@@ -818,57 +817,57 @@ public boolean isEmpty() {
         return idToKeys;
     }
 
-    protected void putAllInternal(final Map<? extends Object, ? extends Object> entries) {
-        final NodeEngine nodeEngine = getNodeEngine();
-        final MapService mapService = getService();
-        int factor = 3;
+    /**
+     * This Operation will first group all puts per partition and then send a PutAllOperation per partition. So if there are e.g.
+     * 5 keys for a single partition, then instead of having 5 remote invocations, there will be only 1 remote invocation.
+     *
+     * If there are multiple puts for different partitions on the same member, they are executed as different remote operations.
+     * Probably this can be optimized in the future by making use of an PartitionIterating operation.
+     *
+     * @param entries
+     */
+    protected void putAllInternal(Map<? extends Object, ? extends Object> entries) {
+        NodeEngine nodeEngine = getNodeEngine();
         InternalPartitionService partitionService = nodeEngine.getPartitionService();
         OperationService operationService = nodeEngine.getOperationService();
         int partitionCount = partitionService.getPartitionCount();
-        boolean tooManyEntries = entries.size() > (partitionCount * factor);
+        MapServiceContext mapServiceContext = getService().getMapServiceContext();
+
         try {
-            if (tooManyEntries) {
-                List<Future> futures = new LinkedList<Future>();
-                Map<Integer, MapEntrySet> entryMap
-                        = new HashMap<Integer, MapEntrySet>(nodeEngine.getPartitionService().getPartitionCount());
-                for (Entry entry : entries.entrySet()) {
-                    checkNotNull(entry.getKey(), NULL_KEY_IS_NOT_ALLOWED);
-                    checkNotNull(entry.getValue(), NULL_VALUE_IS_NOT_ALLOWED);
-
-                    int partitionId = partitionService.getPartitionId(entry.getKey());
-                    if (!entryMap.containsKey(partitionId)) {
-                        entryMap.put(partitionId, new MapEntrySet());
-                    }
-                    entryMap.get(partitionId).add(
-                            new AbstractMap.SimpleImmutableEntry<Data, Data>(mapService.getMapServiceContext().toData(
-                                    entry.getKey(),
-                                    partitionStrategy),
-                                    mapService.getMapServiceContext()
-                                            .toData(entry.getValue())
-                            ));
+            List<Future> futures = new ArrayList<Future>(partitionCount);
+            MapEntrySet[] entrySetPerPartition = new MapEntrySet[partitionCount];
```

Comment:
I foresee this being somewhat expensive where entries.size() << partitionCount (with hazelcast defaults, that's 271 elements), but as entries.size() approaches partitionCount, the HashMap used by the original code would be worse. Ideally we would be able to easily switch between the two depending on how many entries we're putting in, but that's beyond the scope of this PR.